Psychtoolbox/PsychOpenGL/PsychGLSLShaders/ColorGratingShader.frag.txt

```diff
@@ -0,0 +1,59 @@
+/*
+ * File: ColorGratingShader.frag.txt
+ * Shader for drawing of color gratings.
+ *
+ * Copyright 2014, Ian Andolina <http://github.com/iandol>, licenced under the MIT Licence
+ *
+ */
+
+uniform vec2    center;
+uniform vec4    color1;
+uniform vec4    color2;
+uniform float   radius;
+
+varying vec3    baseColor;
+varying float   alpha;
+varying float   phase;
+varying float   frequency;
+varying float   sigma;
+varying float   contrast;
+
+void main() {
+    //current position
+    vec2 pos = gl_TexCoord[0].xy;
+
+    /* find our distance from center, if distance to center (aka radius of pixel) > Radius, discard this pixel: */
+    if ( radius > 0.0 ) {
+        float dist = distance( pos, center );
+        if ( dist > radius ) discard;
+    }
```

Comment:
This is just a remark wrt. to the if radius... section.

Depending on the graphics card, and how modern the graphics driver is - or more specifically how advanced the optimizer of the GLSL to gpu machine language compiler is, it could be more efficient overall to simply do:

if (distance(pos, center) > radius) discard;

and define radius = inf in the CreateProcedural... M-File, instead of radius = 0;

The reason is that many GPU's don't like branching (aka if/switch/while) very much, it reduces efficiency. Sometimes they solve if/else branches by executing the code from both if(true) and if(false) for each pixel fragment, and then just throw away the result from the branch that was the wrong one for a fragment (output masking). So nesting if statements can be bad for performance. If and how much depends on the fancyness of the shader compiler.

What gpu's really like is a linear control flow - or anything very predictable.

I don't know if or how much this would matter for performance in this specific case. Likely not at all or not much in the typical use case for this shader. I just thought i mention it as a general rule of thumb.

Reply:
Thanks!